feat: Editorial review for @scope docs (#30290)
mdn/content: files/en-us/web/css/@scope/index.md

{{CSSRef}}{{SeeCompatTable}}

> The **`@scope`** [CSS](/en-US/docs/Web/CSS) [at-rule](/en-US/docs/Web/CSS/At-rule) enables you to select elements in specific DOM subtrees, targetting elements precisely without writing overly specific selectors that are hard to override, and without coupling your selectors too tightly to the DOM structure.

Reviewer: ```suggestion
The **`@scope`** [CSS](/en-US/docs/Web/CSS) [at-rule](/en-US/docs/Web/CSS/At-rule) enables you to select elements in specific DOM subtrees, targeting elements precisely without writing overly-specific selectors that are hard to override, and without coupling your selectors too tightly to the DOM structure.
```

In JavaScript, `@scope` can be accessed via the CSS object model interface {{DOMxRef("CSSScopeRule")}}.

## Syntax

The `@scope` at-rule contains one or more rulesets (termed **scoped style rules**) and defines a scope in which to apply them to selected elements. `@scope` can be used in two ways:

1. As a standalone block inside your CSS, in which case it includes a prelude section that includes **scope root** and optional **scope limit** selectors — these define the upper and lower bounds of the scope.

   ```css
   @scope (scope root) to (scope limit) {
     rulesets
   }
   ```

2. As inline styles included inside a {{htmlelement("style")}} element in your HTML, in which case the prelude is omitted, and the enclosed ruleset is automatically scoped to the `<style>` element's enclosing parent element.

   ```html
   <parent-element>
     <style>
       @scope {
         rulesets
       }
     </style>
   </parent-element>
   ```

## Description

> A complex web document tends to include several distinct components — from headers, footers, and new articles, to maps, media players, and ads. As complexity increases, it becomes more important to manage the styling for these components effectively, and effective scoping of styles is a key consideration. Let's consider the following DOM tree:

Reviewer: ```suggestion
A complex web document might include components such as headers, footers, news articles, maps, media players, ads, and others. As complexity increases, effectively managing the styling for these components becomes more of a concern, and effective scoping of styles helps us manage this complexity. Let's consider the following DOM tree:
```

```plain-nolint
body
└─ article.feature
   ├─ section.article-hero
   │  ├─ h2
   │  └─ img
   │
   ├─ section.article-body
   │  ├─ h3
   │  ├─ p
   │  ├─ img
   │  ├─ p
   │  └─ figure
   │     ├─ img
   │     └─ figcaption
   │
   └─ footer
      ├─ p
      └─ img
```

> If we wanted to select just the {{htmlelement("img")}} element inside the {{htmlelement("section")}} with a class of `article-body`, we could:

Reviewer: ```suggestion
If we wanted to select the {{htmlelement("img")}} element inside the {{htmlelement("section")}} with a class of `article-body`, we could do the following:
```

- Write a selector like `.feature > .article-body > img`. However, that has high specificity so is hard to override, and is also tighly coupled to the DOM structure. If your markup structure changes in the future, you might need to rewrite your CSS.
- Write something less specific like `.article-body img`. However, that will select all images inside the `section`.

This is where `@scope` is useful. It allows you to define a precise scope inside which your selectors are allowed to target elements. For example, we could solve the above problem using a standalone `@scope` block like the following:

```css
@scope (.article-body) to (figure) {
  img {
    border: 5px solid black;
    background-color: goldenrod;
  }
}
```

The `.article-body` scope root selector defines the upper bound of the DOM tree scope in which the ruleset will be applied, and the `figure` scope limit selector defines the lower bound. As a result, only {{htmlelement("img")}} elements inside a `<section>` with a class of `article-body`, but not inside {{htmlelement("figure")}} elements, will be selected.

> > **Note:** This kind of scoping — with an upper and lower bound — is commonly referred to as a **donut scope**.

Reviewer: 🍩

    background-color: goldenrod;
  }
}
```

Or you could include your `@scope` block inline inside a `<style>` element, which in turn is inside the `<section>` with a class of `article-body`:

```html
<section class="article-body">
  <style>
    @scope {
      img {
        border: 5px solid black;
        background-color: goldenrod;
      }
    }
  </style>

  <!-- ... -->
</section>
```

> **Note:** It is important to understand that, while `@scope` allows you to isolate the application of selectors to specific DOM subtrees, it does not completely isolate the applied styles to within those subtrees. This is most noticeable with inheritance — properties that are inherited by children (for example {{cssxref("color")}} or {{cssxref("font-family")}}) will still be inherited, beyond any set scope limit.

### The `:scope` pseudo-class

In the context of a `@scope` block, the {{cssxref(":scope")}} pseudo-class represents the scope root — it provides an easy way to apply styles to the scope root itself, from inside the scope:

```css
@scope (.feature) {
  :scope {
    background: rebeccapurple;
    color: antiquewhite;
    font-family: sans-serif;
  }
}
```

> `:scope` is implicitly prepended to all scoped style rules. If you wish to, you can write it explicitly, or you could prepend the [nesting](/en-US/docs/Web/CSS/CSS_nesting) selector (`&`) to get the same effect. (You might find these styles easier to understand.)
> 
> As such, the three lines in the following block are all equivalent in what they select:

Reviewer: ```suggestion
In fact, `:scope` is implicitly prepended to all scoped style rules. If you want, you can explicitly include it before scoped rules or prepend the [nesting](/en-US/docs/Web/CSS/CSS_nesting) selector (`&`) to get the same effect if you find these styles easier to understand.

For illustration, the three rules in the following block are all equivalent in what they select:
```

Author: I mostly agreed with this but made a few tweaks in my next PR.

- It is perfectly valid to define the scope root and limit as a selector list, in which case multiple scopes will be defined. In the following example the styles are applied to any `<img>` inside a `<section>` with a class of `article-hero` or `article-body`, but not if it is nested inside a `<figure>`:

  ```css
  @scope (.article-hero, .article-body) to (figure) {
    img {
      border: 5px solid black;
      background-color: goldenrod;
    }
  }
  ```

### Specificity in `@scope`

Including a ruleset inside a `@scope` block does not affect the specificity of its selector, regardless of the selectors used inside the scope root and limit. For example:

```css
@scope (.article-body) {
  /* img has a specificity of 0-0-1, as expected */
  img { ... }
}
```

However, if you decide to explicitly prepend the `:scope` pseudo-class to your scoped selectors, you'll need to factor it in when calculating their specificity. `:scope`, like all regular pseudo-classes, has a specificity of 0-1-0. For example:

```css
@scope (.article-body) {
  /* :scope img has a specificity of 0-1-0 + 0-0-1 = 0-1-1 */
  :scope img { ... }
}
```

> When using the `&` selector inside a `@scope` block, `&` represents the scope root selector; it is internally rewritten to that selector wrapped inside an {{cssxref(":is", ":is()")}} pseudo-class function. So for example, in:

Reviewer: what do you think about this one?

```suggestion
When using the `&` selector inside a `@scope` block, `&` represents the scope root selector; it is internally calculated as that selector wrapped inside an {{cssxref(":is", ":is()")}} pseudo-class function. So for example, in:
```

```css
@scope (figure, #primary) {
  & img { ... }
}
```

> `& img` is equivalent to `:is(figure, #primary) img`.
> 
> Since `:is()` takes the specificity of its most specific argument (`#primary`, in this case), the specificity of the scoped `& img` selector is therefore 1-0-0 + 0-0-1 = 1-0-1.

Reviewer: Feel free to dismiss, but I find the line break stops the flow a little.

```suggestion
`& img` is equivalent to `:is(figure, #primary) img`.
Since `:is()` takes the specificity of its most specific argument (`#primary`, in this case), the specificity of the scoped `& img` selector is therefore 1-0-0 + 0-0-1 = 1-0-1.
```

Author: I decided to check the rendering out first locally, but in the end, I agreed and made them into one paragraph.

### The difference between `:scope` and `&` inside `@scope`

> `:scope` represents the matched scope root, whereas `&` represents the selector used to match the scope root.
> 
> Because of this, it is possible to chain `&` multiple times, whereas you can only use `:scope` once; as you can't match a scope root inside a scope root.

Reviewer: Same [as above](https://github.com/mdn/content/pull/30290/files#r1401184002)

```suggestion
`:scope` represents the matched scope root, whereas `&` represents the selector used to match the scope root.
Because of this, it is possible to chain `&` multiple times, whereas you can only use `:scope` once; as you can't match a scope root inside a scope root.
```

Author: I agreed with this in the end too, but made a few improvements to the prose as well as merging the two paras into one.

```css
@scope (.feature) {
  /* Selects a `.feature` inside the matched root .feature */
  & & { ... }

  /* Doesn't work */
  :root :root { ... }
}
```

### How `@scope` conflicts are resolved

> `@scope` adds a new criterion to the [CSS cascade](/en-US/docs/Web/CSS/CSS_cascade): **scoping proximity**. This states that when two scopes have conflicting styles, the style that has the smallest number of hops up the DOM tree hierarchy to the scope root wins. Let's look at an example to see what this means.

Reviewer: Maybe "is applied" is better.
 
```suggestion
`@scope` adds a new criterion to the [CSS cascade](/en-US/docs/Web/CSS/CSS_cascade): **scoping proximity**. This states that when two scopes have conflicting styles, the style that has the smallest number of hops up the DOM tree hierarchy to the scope root is applied. Let's look at an example to see what this means.
```

Take the following HTML snippet, where we are nesting different-themed cards inside one another:

```html
<div class="light-theme">
  <p>Light theme text</p>
  <div class="dark-theme">
    <p>Dark theme text</p>
    <div class="light-theme">
      <p>Light theme text</p>
    </div>
  </div>
</div>
```

If we wrote the theme CSS like so, we'd run into trouble:

```css
.light-theme {
  background: #ccc;
}

.dark-theme {
  background: #333;
}

.light-theme p {
  color: black;
}

.dark-theme p {
  color: white;
}
```

> The innermost paragraph is supposed to be colored black because it is in a light theme card. However, it is being targetted by both `.light-theme p` and `.dark-theme p` and, since the `.dark-theme p` rule appears later in the source order, it wins and the paragraph ends up being wrongly colored white.

Reviewer: What do you think?

```suggestion
The innermost paragraph is supposed to be colored black because it is in a light theme card. However, it's targeted by both `.light-theme p` and `.dark-theme p`. The `.dark-theme p` rule appears later in the source order; it's applied, and the paragraph is unintentionally colored white.
```

Author: I used some of this, but I kept "wrongly"as I felt it captured the intent fine.

  <section class="article-hero">
    <h2>Article heading</h2>
    <img alt="image" />
  </section>

  <section class="article-body">
    <h3>Article subheading</h3>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam euismod
      consectetur leo, nec eleifend quam volutpat vitae. Duis quis felis at
      augue imperdiet aliquam. Morbi at felis et massa mattis lacinia. Cras
      pharetra velit nisi, ac efficitur magna luctus nec.
    </p>

    <img alt="image" />

    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>

    <figure>
      <img alt="image" />
      <figcaption>My infographic</figcaption>
    </figure>
  </section>

  <footer>
    <p>Written by Chris Mills.</p>
    <img alt="image" />
  </footer>
</article>
```

#### CSS

```css hidden
* {
  box-sizing: border-box;
}

article {
  margin-bottom: 20px;
  padding: 10px;
  border: 2px solid gray;
}

.article-hero,
.article-body,
article footer {
  padding: 20px;
  margin: 10px;
  border: 2px solid lightgray;
}

img {
  height: 100px;
  width: 100%;
  display: block;
  background-color: lightgray;
  color: black;
  padding: 10px;
}
```

In our CSS, we have two `@scope` blocks:

> - The first one defines its scope root as elements with a class of `.feature` (in this case, the outer `<div>` only), again demonstrating how `@scope` can be used to theme a specific HTML subset.

Reviewer: ```suggestion
- The first `@scope` block defines its scope root as elements with a class of `.feature` (in this case, the outer `<div>` only), demonstrating how `@scope` can be used to theme a specific HTML subset.
```

> - The second one again defines its scope root as elements with a class of `.feature`, but also defines a scope limit of `figure`. This ensures that contained rulesets will only be applied to matching elements within the scope root (`<div class="figure"> ... </div>` in this case) that ARE NOT nested inside descendant `<figure>` elements. This `@scope` block contains a single ruleset that styles `<img>` elements with a thick black border and a golden background color.

Reviewer: ```suggestion
- The second `@scope` block also defines its scope root as elements with a class of `.feature`, but defines a scope limit of `figure`. This ensures that contained rulesets will only be applied to matching elements within the scope root (`<div class="figure"> ... </div>` in this case) that **are not** nested inside descendant `<figure>` elements. This `@scope` block contains a single ruleset that styles `<img>` elements with a thick black border and a golden background color.
```

Author: mostly used, with a little tweaking.
